web/src/main/java/se/inera/intyg/intygstjanst/web/service/impl/CertificateEventRevokeServiceImpl.java

```diff
@@ -0,0 +1,147 @@
+import org.springframework.stereotype.Service;
+import se.inera.intyg.common.support.xml.XmlMarshallerHelper;
+import se.inera.intyg.intygstjanst.web.exception.RecipientUnknownException;
+import se.inera.intyg.intygstjanst.web.service.CertificateEventRevokeService;
+import se.inera.intyg.intygstjanst.web.service.MonitoringLogService;
+import se.inera.intyg.intygstjanst.web.service.RecipientService;
+import se.inera.intyg.intygstjanst.web.service.dto.GetCertificateXmlResponse;
+import se.inera.intyg.intygstjanst.web.service.dto.RevokedInformationDTO;
+import se.inera.intyg.intygstjanst.web.service.dto.StaffDTO;
+import se.riv.clinicalprocess.healthcond.certificate.revokeCertificate.v2.RevokeCertificateResponderInterface;
+import se.riv.clinicalprocess.healthcond.certificate.revokeCertificate.v2.RevokeCertificateResponseType;
+import se.riv.clinicalprocess.healthcond.certificate.revokeCertificate.v2.RevokeCertificateType;
+import se.riv.clinicalprocess.healthcond.certificate.types.v3.HsaId;
+import se.riv.clinicalprocess.healthcond.certificate.v3.Enhet;
+import se.riv.clinicalprocess.healthcond.certificate.v3.HosPersonal;
+import se.riv.clinicalprocess.healthcond.certificate.v3.ResultCodeType;
+
+@Service
+@Slf4j
+public class CertificateEventRevokeServiceImpl implements CertificateEventRevokeService {
+
+    private static final String HSA_ID_OID = "1.2.752.129.2.1.4.1";
+
+    private final RevokeCertificateResponderInterface revokeCertificateResponderInterface;
+    private final RecipientService recipientService;
+    private final MonitoringLogService monitoringLogService;
+
+    public CertificateEventRevokeServiceImpl(
+        @Qualifier("revokeCertificateClient") RevokeCertificateResponderInterface revokeCertificateResponderInterface,
+        RecipientService recipientService, MonitoringLogService monitoringLogService) {
+        this.revokeCertificateResponderInterface = revokeCertificateResponderInterface;
+        this.recipientService = recipientService;
+        this.monitoringLogService = monitoringLogService;
+    }
+
+    @Override
+    public void revoke(GetCertificateXmlResponse xmlResponse, String decodedXml) {
+        try {
+            final var logicalAddress = recipientService.getRecipient(xmlResponse.getRecipient().getId()).getLogicalAddress();
+            final var request = getRequest(xmlResponse.getRevoked(), xmlResponse.getUnitId(), decodedXml);
+            final var wsResponse = revokeCertificateResponderInterface.revokeCertificate(logicalAddress, request);
+            handleResponse(wsResponse, xmlResponse);
+        } catch (SOAPFaultException | RecipientUnknownException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private static RevokeCertificateType getRequest(RevokedInformationDTO revokedInformation, String unitId,
+        String decodedXml) {
+        final var jaxbElement = XmlMarshallerHelper.unmarshal(decodedXml);
```

Comment:
This xml should not be part of the RevokeCertificateType. You should create a new RevokeCertificateType and set necessary information. Check IntygServiceImpl.revokeIntyg in Webcert for more information on how it is done today.

Reply:
Done